src/main/java/graphql/validation/rules/UniqueVariableNamesRule.java

```diff
@@ -0,0 +1,47 @@
+package graphql.validation.rules;
+
+import graphql.language.*;
+import graphql.validation.AbstractRule;
+import graphql.validation.ValidationContext;
+import graphql.validation.ValidationErrorCollector;
+import graphql.validation.ValidationErrorType;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Unique variable names
+ * <p>
+ * A GraphQL operation is only valid if all its variables are uniquely named.
+ */
+public class UniqueVariableNamesRule extends AbstractRule {
+
+    public UniqueVariableNamesRule(ValidationContext validationContext, ValidationErrorCollector validationErrorCollector) {
+        super(validationContext, validationErrorCollector);
+    }
+
+    @Override
+    public void checkOperationDefinition(OperationDefinition operationDefinition) {
+        List<VariableDefinition> variableDefinitions = operationDefinition.getVariableDefinitions();
+        if (variableDefinitions == null || variableDefinitions.size() <= 1) {
+            return;
+        }
+
+        Set<String> variableNameList = new HashSet<>(variableDefinitions.size());
```

Comment:
we use `LinkedHashSet` in general in favor of `HashSet` because it preserves the order.